cajita/src/Cajita_SparseIndexSpace.hpp

```diff
@@ -0,0 +1,499 @@
+    };
+    KOKKOS_FORCEINLINE_FUNCTION
+    constexpr auto operator()( std::array<int, 3> tileid ) -> Key
+    {
+        return bit_spread( page_zmask, tileid[2] ) |
+               bit_spread( page_ymask, tileid[1] ) |
+               bit_spread( page_xmask, tileid[0] );
+    }
+
+    KOKKOS_FORCEINLINE_FUNCTION
+    constexpr auto operator()( int tileid_x, int tileid_y, int tileid_z ) -> Key
+    {
+        return bit_spread( page_zmask, tileid_z ) |
+               bit_spread( page_ymask, tileid_y ) |
+               bit_spread( page_xmask, tileid_x );
+    }
+};
+
+// can be rewriten in a recursive way
+template <typename Key>
+struct HashKey2TileID<Key, HashTypes::Naive>
+{
+    HashKey2TileID( std::array<int, 3> &&tnum )
+        : _tilenum( tnum )
+    {
+    }
+    HashKey2TileID( int i, int j, int k )
+        : _tilenum( i, j, k )
+    {
+    }
+
+    KOKKOS_FORCEINLINE_FUNCTION
+    constexpr void operator()( Key tilekey, std::array<int, 3> &tileid )
+    {
+        tileid[2] = tilekey % _tilenum[2];
+        tileid[1] = static_cast<Key>( tilekey / _tilenum[2] ) % _tilenum[1];
+        tileid[0] = static_cast<Key>( tilekey / _tilenum[2] / _tilenum[1] ) %
+                    _tilenum[0];
+    }
+
+    KOKKOS_FORCEINLINE_FUNCTION
+    constexpr void operator()( Key tilekey, int &tileid_x, int &tileid_y,
+                               int &tileid_z )
+    {
+        tileid_z = tilekey % _tilenum[2];
+        tileid_y = static_cast<Key>( tilekey / _tilenum[2] ) % _tilenum[1];
+        tileid_x = static_cast<Key>( tilekey / _tilenum[2] / _tilenum[1] ) %
+                   _tilenum[0];
+    }
+
+  private:
+    std::array<int, 3> _tilenum;
+};
+
+template <typename Key>
+struct HashKey2TileID<Key, HashTypes::Morton>
+{
+    HashKey2TileID( std::array<int, 3> && ) {} // ugly
+    HashKey2TileID( int, int, int ) {}
+    enum : uint64_t
+    { // hand-coded now, can be improved by iterative func
+        page_zmask = ( 0x9249249249249249UL ),
+        page_ymask = ( 0x2492492492492492UL ),
+        page_xmask = ( 0x4924924924924924UL )
+    };
+
+    KOKKOS_FORCEINLINE_FUNCTION
+    constexpr void operator()( Key tilekey, std::array<int, 3> &tileid )
+    {
+        tileid[2] = bit_pack( page_zmask, tilekey );
+        tileid[1] = bit_pack( page_ymask, tilekey );
+        tileid[0] = bit_pack( page_xmask, tilekey );
+    }
+
+    KOKKOS_FORCEINLINE_FUNCTION
+    constexpr void operator()( Key tilekey, int &tileid_x, int &tileid_y,
+                               int &tileid_z )
+    {
+        tileid_z = bit_pack( page_zmask, tilekey );
+        tileid_y = bit_pack( page_ymask, tilekey );
+        tileid_x = bit_pack( page_xmask, tilekey );
+    }
+};
+
+//---------------------------------------------------------------------------//
+/*!
+  \class SparseIndexSpace
+  \brief Sparse index space, hierarchical structure (cell->tile->block)
+  \ ValueType : tileNo type
+ */
+template <int N = 3, unsigned long long TileNPerDim = 4,
```

Comment:
I did consider this part. I'm trying to fit the design as the index space of uniform grids. Here I only considered 3D cases, with the first three for the `ijk`, the last one, if there is, for the channel. That is, only N=3 or 4 are possible. I actually don't think its a generalizable design and also thinking about changing it. Because if the channel dimension involves, I'm not sure how to identify if the spatial dimension is 2 or 3 when N = 3.

The mapping between ijk and id number is hardcoded for 3D now, I can change them to recursive implementation for support for both 2D and 3D in the next round refactor.

Reply:
That is fine - we can overload for 2d later. No need to do recursive implementation. I think the bit operations will give better performance and work for what we want.